File: idakonyvtar/src/main/java/eu/eyan/idakonyvtar/model/Book.java
package eu.eyan.idakonyvtar.model;

import static com.google.common.collect.Lists.newArrayList;

import java.beans.PropertyChangeEvent;
import java.util.Collections;
import java.util.List;

import lombok.Getter;

import com.jgoodies.binding.beans.Model;

public class Book extends Model implements Cloneable
{
    private static final long serialVersionUID = 1L;
    final List<String> values = newArrayList();

    public Book(int columnSzám)
    {
        // TODO az arraylist.set metódus miatt :(
        for (int i = 0; i < columnSzám; i++)
        {
            values.add("");
        }
    }

    public Book(Book book)
    {
        this(book.values.size());
        Collections.copy(this.values, book.values);
    }

    public void setValue(int columnIndex, final String value)
    {
        String oldValue = this.values.get(columnIndex);
        // FIXME: AutoCompleteDecorator Problem: Disgusting Hack but works...
        if (value != null /* && !value.equals("") */)
        {
            this.values.set(columnIndex, value);
            firePropertyChange(new BookPropertyChangeEvent(this, "PROP", oldValue, value, columnIndex));
        }
    }

    public String getValue(int index)
    {
        return values.get(index);
    }

    public static class BookPropertyChangeEvent extends PropertyChangeEvent
    {
        private static final long serialVersionUID = 1L;
        @Getter
        private int columnIndex;

        public BookPropertyChangeEvent(Object source, String propertyName, Object oldValue, Object newValue, int columnIndex)
        {
            super(source, propertyName, oldValue, newValue);
            this.columnIndex = columnIndex;
        }
    }

    public static class Builder
    {
        private Book book;

        public Builder(int columnCount)
        {
            this.book = new Book(columnCount);
        }

        public Builder withValue(int columnIndex, String value)
        {
            this.book.setValue(columnIndex, value);
            return this;
        }

        public Book build()
        {
            return this.book;
        }
    }
}

File: idakonyvtar/src/main/scala/eu/eyan/idakonyvtar/model/Book.scala
package eu.eyan.idakonyvtar.model;

import java.beans.PropertyChangeEvent

import scala.collection.mutable.MutableList

import com.jgoodies.binding.beans.Model
import eu.eyan.log.Log
import eu.eyan.log.Log
import java.awt.Image
import scala.collection.mutable.Map
import java.awt.image.BufferedImage
import eu.eyan.idakonyvtar.text.TechnicalTextsIda._
import eu.eyan.idakonyvtar.oszk.Marc
import eu.eyan.util.excel.ExcelColumn

trait BookFieldType
case object Multifield extends BookFieldType
case object Autocomplete extends BookFieldType
case object InTable extends BookFieldType
case object Remember extends BookFieldType
case object Picture extends BookFieldType

case class BookField (excelColumn: ExcelColumn, fieldName: String, private val fieldTypes : List[BookFieldType], val marcCodes: Array[Marc]) {
  lazy val isMulti = fieldTypes.contains(Multifield)
  lazy val isAutocomplete = fieldTypes.contains(Autocomplete)
  lazy val isShowInTable = fieldTypes.contains(InTable)
  lazy val isPicture = fieldTypes.contains(Picture)
  lazy val isRemember = fieldTypes.contains(Remember)
}

class BookPropertyChangeEvent(source: Object, propertyName: String, oldValue: Object, newValue: Object, val field: BookField) extends PropertyChangeEvent(source, propertyName, oldValue, newValue)

object Book {
  def apply(fields: List[(BookField, String)]): Book = new Book(Map(fields: _*))

  def empty(fields: List[BookField]): Book = Book(fields.map((_, EMPTY_STRING)))

  def copy(book: Book): Book = book.copy
}

// TODO: make it immutable. remove the binding framework
class Book private (
  private val fields: Map[BookField, String]        = Map(),
  private val images: Map[BookField, BufferedImage] = Map()) extends Model {

  def setValue(field: BookField)(value: String) = {
    val oldValue: String = getValue(field)
    fields.put(field, value)
    Log.trace("fire " + value)
    firePropertyChange(new BookPropertyChangeEvent(this, "PROP", oldValue, value, field))
  }

  def getValue(field: BookField) = fields.get(field).getOrElse("") // TODO is getOrElse ok??? // it was done only for tests

  def setImage(field: BookField)(image: BufferedImage) = images.put(field, image)
  def getImage(field: BookField) = images.get(field)

  override def toString(): String = fields.mkString(" - ")

  protected def copy = new Book(fields.clone, images.clone)
}